src/lib/backoff.ts

```diff
@@ -0,0 +1,88 @@
+import { promisify } from 'util';
+
+const sleep = promisify(setTimeout);
+
+export type BackoffCallback = {
+	retryCount: number;
+	delay: number;
+	error: any;
+};
+
+export type Options = {
+	attempt: number;
+	maxDelay: number;
+	minDelay: number;
+	maxRetries: number;
+	retryCallback?: (info: BackoffCallback) => void;
+};
+
+const DEFAULT_OPTIONS: Partial<Options> = {
+	attempt: 0,
+	maxDelay: 15000,
+	minDelay: 15000,
+	maxRetries: Infinity,
+};
+
+/**
+ * Retries a function with exponential delay between calls.
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ *
+ * Supports exceptions which return `retryAfter` value to specify the backoff duration.
+ * Will call `retryCallback` if passed in options as a way for callers to react to retries
+ */
+export function withBackoff<T extends (...args: any[]) => any>(
+	func: T,
+	options: Partial<Options> = {},
+) {
+	// If the function returns a promise, unwrap the promise return type
+	// otherwise use the actual return
+	type TReturn = ReturnType<T> extends Promise<infer R> ? R : ReturnType<T>;
+
+	const normalizedOptions: Options = {
+		...DEFAULT_OPTIONS,
+		...options,
+	} as Options;
+
+	return async function wrapped(...args: Parameters<T>): Promise<TReturn> {
+		let error: any;
+		try {
+			return await func(...args);
+		} catch (e) {
+			error = e;
+			if (e.retryAfter) {
+				normalizedOptions.attempt = 0;
+				normalizedOptions.minDelay = e.retryAfter;
```

Comment:
if `maxDelay < e.retryAfter` this won't have the intended effect.

Reply:
good point...the delay does not really consider that the value is from the exception which should be respected regardless of what the min and max are. I will modify the delay to use min and max but retryAfter over rules them all.

Reply:
I removed setting minDelay to retryAfter because I just set the wait time equal to the retryAfter value. I am still setting attempt = 0 so if the API stops sending retryAfter the exponential curve does not resume a previously high value.